firebase_admin/db.py

```diff
@@ -67,6 +78,70 @@ def _parse_path(path):
             'Invalid path: "{0}". Path contains illegal characters.'.format(path))
     return [seg for seg in path.split('/') if seg]
 
+class KeepAuthSession(Session):
+    """
+    A session that doesn't drop Authentication on redirects between domains.
+    """
+
+    def rebuild_auth(self, prepared_request, response):
+        pass
+
+class ClosableSSEClient(SSEClient):
+    def __init__(self, *args, **kwargs):
+        self.should_connect = True
+        super(ClosableSSEClient, self).__init__(*args, **kwargs)
+
+    def _connect(self):
+        if self.should_connect:
+            super(ClosableSSEClient, self)._connect()
+        else:
+            raise StopIteration()
+
+    def close(self):
+        self.should_connect = False
+        self.retry = 0
+        self.resp.raw._fp.fp.raw._sock.shutdown(socket.SHUT_RDWR)
+        self.resp.raw._fp.fp.raw._sock.close()
+
+class Stream:
+    def __init__(self, url, stream_handler, build_headers, stream_id):
+        self.build_headers = build_headers
+        self.url = url
+        self.stream_handler = stream_handler
+        self.stream_id = stream_id
+        self.sse = None
+        self.thread = None
+        self.start()
+
+    def make_session(self):
+        """
+        Return a custom session object to be passed to the ClosableSSEClient.
+        """
+        session = KeepAuthSession()
+        return session
+
+    def start(self):
+        self.thread = threading.Thread(target=self.start_stream)
+        self.thread.start()
+        return self
+
+    def start_stream(self):
+        self.sse = ClosableSSEClient(self.url, session=self.make_session(), build_headers=self.build_headers)
```

Comment:
This causes an error if pyrebase is not installed.

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.5/site-packages/firebase_admin/db.py", line 129, in start_stream
    self.sse = ClosableSSEClient(self.url, session=self.make_session(), build_headers=self.build_headers)
  File "/usr/local/lib/python3.5/site-packages/firebase_admin/db.py", line 88, in __init__
    super(ClosableSSEClient, self).__init__(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/sseclient.py", line 39, in __init__
    self._connect()
  File "/usr/local/lib/python3.5/site-packages/firebase_admin/db.py", line 94, in _connect
    super(ClosableSSEClient, self)._connect()
  File "/usr/local/lib/python3.5/site-packages/sseclient.py", line 47, in _connect
    self.resp = requester.get(self.url, stream=True, **self.requests_kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 521, in get
    return self.request('GET', url, **kwargs)
TypeError: request() got an unexpected keyword argument 'build_headers'
```

Reply:
Let me check. I'll get back to you on this.

Reply:
Hi - Any luck with this one, I am getting exactly the same error. What is the fix?

Reply:
Awesome. Thanks. Waiting for the patch. By the way. This is an awesome library.

Reply:
Hi - Any luck?

Reply:
Hey @vijjuk 
Sorry for the delay
Can you kindly give this one a try, it's working at my end. 
link:   https://github.com/Aqsa-K/firebase-admin-python

Reply:
@vijjuk 
You cannot pip update this right now. You'll have to git pull it from here: https://github.com/Aqsa-K/firebase-admin-python
Once you have pulled the changes, remember to run setup.py as follows
- python setup.py build
- python setup.py install

Reply:
@darrenswhite, can you give the updated version a try? 
https://github.com/Aqsa-K/firebase-admin-python

Reply:
@vijjuk 
Yes firebase-admin is much better. I shifted from pyrebase to firebase-admin. 
firebase-admin is using sse-client library to listen for changes in firebase database.

Reply:
Hello, can I ask what is the status with this pull request? Is it going to be merged? I am developing an app, that requires the server to listen to changes in the firebase database. @Aqsa-K  I tried your fork repo, somehow it fires off an event as soon as I set the `stream()` and the callback function is called immediately. But the intended listening event is triggered every time when the data is changed in the database side. 

```
>>> ref.stream(callbackfunc)
<firebase_admin.db.Stream object at 0x10c9b7ef0>
>>> {'event': 'put', 'data': True, 'path': '/'}
>>> # The event above is fired off automatically on stream without modification on db

>>> # these 3 events below are fired off by the changes in the db. 
>>> {'event': 'put', 'data': 'true', 'path': '/'}
>>> {'event': 'put', 'data': 'false', 'path': '/'}
>>> {'event': 'put', 'data': 'true', 'path': '/'}

>>> self.db.reference("/users/dev1/successful").stream(callbackfunc)
<firebase_admin.db.Stream object at 0x10ef897f0>
>>> {'event': 'put', 'data': True, 'path': '/'}
>>> # The event above is fired off automatically on stream without modification on db
```

Is it possible to fix this unexpected behavior? I would like to help fix it, but I have no idea where to start looking and I have never worked with SSE before. If anyone can point me in correct direction, please do.

Reply:
The first event is fired initially to retrieve all of the current data in the database.

Reply:
@the-c0d3r you can separate out the first event which is fired initially by filtering the type 'event' or the type 'path' in the message received by callback function. 

If you intend to look for changes in your data, then that should have the type 'event' as 'put'.

Let me know if this is still unclear and you need more help with this.

Reply:
Hi, @Aqsa-K Thanks for your comment. I have implemented functionality to filter the events and to only fire them if the value changes. I also created [a PR](https://github.com/Aqsa-K/firebase-admin-python/pull/1) on your fork. I noticed a few things on the SSE client. When I close the stream, it takes about 10 to 20 seconds for it to return. The following close function triggers some attribute error as well. So I catch them with an exception. 

```python
    def close(self):
        self.should_connect = False
        self.retry = 0
        try:
            self.resp.close()
            self.resp.raw._fp.fp.raw._sock.shutdown(socket.SHUT_RDWR)
            self.resp.raw._fp.fp.raw._sock.close()
        except AttributeError:
            pass
```

```python
    def close(self):
        while not self.sse and not hasattr(self.sse, 'resp'):
            time.sleep(0.001)
        self.sse.running = False
        self.sse.close()
        self.thread.join()
        #  return self
```

But the problem still remains, the closing of the stream takes too long for a single threaded operation for my app. The `close()` function of the stream waits for the sse client thread to join. I think that takes up the most time. Is there any way to force the closing of the stream? I need to watch the changes based on the users' query (which differs everytime), then after getting results, I have to dispose of the stream. One workaround I can think of is to do that in a threadpool, so the thread can wait on the sse client closing process, but that does not seem to be the optimal way to me. Or am I doing anything wrong by trying to close?

The following is the way I dispose of the streams

```
streamRef = db.reference("/path/to/obj").stream(lambda x: print(x))
# attempt some changes on the path
streamRef.close()
# takes from 10 to 20 seconds here to return
```

Thanks for reading.